### PR TITLE
[SYCL] Add oneDNN primitive support

### PR DESCRIPTION
**Pull request by luoyu-intel**
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

oneMKL and oneDNN almost share the same GPU kernels, but oneDNN has better compatibility.

llama-bench on A750:
```
DNNL F16
| model                          |       size |     params | backend    | ngl |          test |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------------: | ---------------: |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | SYCL       |  99 |        pp1024 |   1607.21 ± 5.46 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | SYCL       |  99 |          tg32 |     32.01 ± 0.09 |

MKL F16
| model                          |       size |     params | backend    | ngl |          test |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------------: | ---------------: |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | SYCL       |  99 |        pp1024 |   1581.72 ± 4.02 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | SYCL       |  99 |          tg32 |     32.08 ± 0.13 |

DNNL F32
| model                          |       size |     params | backend    | ngl | threads |    sm |          test |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ----: | ------------: | ---------------: |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | SYCL       |  33 |       8 |  none |        pp1024 |    710.00 ± 0.42 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | SYCL       |  33 |       8 |  none |          tg32 |     35.77 ± 0.02 |

MKL F32
| model                          |       size |     params | backend    | ngl | threads |    sm |          test |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ----: | ------------: | ---------------: |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | SYCL       |  33 |       8 |  none |        pp1024 |    709.28 ± 0.92 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | SYCL       |  33 |       8 |  none |          tg32 |     35.74 ± 0.04 |
```

on A770m, the master branch crashed with GGML_SYCL_F16=ON, this branch with oneDNN can run it.
```
DNNL F16

| model                          |       size |     params | backend    | ngl | threads |    sm |          test |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ----: | ------------: | ---------------: |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | SYCL       |  33 |       8 |  none |        pp1024 |   1598.31 ± 1.73 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | SYCL       |  33 |       8 |  none |         tg128 |     40.20 ± 0.09 |
```

Ultra-155H, MKL throws this error: 
```
Native API failed. Native API returns: -2 (PI_ERROR_DEVICE_NOT_AVAILABLE) -2 (PI_ERROR_DEVICE_NOT_AVAILABLE)
Exception caught at file:C:/Users/luoyu/Documents/repo/llama-fork/ggml/src/ggml-sycl.cpp, line:2556, func:operator()
SYCL error: CHECK_TRY_ERROR(dpct::gemm( *stream, oneapi::mkl::transpose::trans, oneapi::mkl::transpose::nontrans, row_diff, src1_ncols, ne10, &alpha_f16, src0_ptr, dpct::library_data_t::real_half, ne00, src1_ptr, dpct::library_data_t::real_half, ne10, &beta_f16, dst_f16.get(), dpct::library_data_t::real_half, ldc, dpct::library_data_t::real_half)): Meet error in this line code!
```
```
DNNL F16
| model                          |       size |     params | backend    | ngl | threads |          test |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ------------: | ---------------: |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | SYCL       |  33 |       8 |        pp1024 |   309.47 ± 39.50 |
| llama 7B Q4_0                  |   3.56 GiB |     6.74 B | SYCL       |  33 |       8 |          tg32 |     11.67 ± 1.07 |
```

### CODE REVIEW

File: ggml/src/CMakeLists.txt [hunk 1]
```diff
@@ -549,6 +549,13 @@ if (GGML_SYCL)
     file(GLOB   GGML_SOURCES_SYCL "ggml-sycl/*.cpp")
     list(APPEND GGML_SOURCES_SYCL "ggml-sycl.cpp")
 
+    find_package(DNNL)
+    message("-- DNNL found:"${DNNL_FOUND})
+    if (GGML_SYCL_TARGET STREQUAL "INTEL")
+        add_compile_definitions(GGML_SYCL_DNNL=${DNNL_FOUND})
+    else()
+        add_compile_definitions(GGML_SYCL_DNNL=0)
+    endif()
```
Comment (airMeng) [inline ggml/src/CMakeLists.txt:558]: OneDNN should also work on other vendors' device

Comment (luoyu-intel) [reply ggml/src/CMakeLists.txt:558]: Others can validate onednn on other platforms, and create a new PR to open it.

Comment (Alcpz) [reply ggml/src/CMakeLists.txt:558]: @airMeng We need to see how oneDNN performs on other platforms first, as the row `gemm` implementation for oneMKL interface introduced synchronization issues for us in the past that needed to be fixed. We should probably switch to oneDNN as well, but we need to ensure that it does not introduce regressions. 

@joeatodd, I agree with @luoyu-intel  here. What do you think?

**Review by airMeng — APPROVED** [review]

**Review by abhilash1910 — APPROVED** [review]

**Review by NeoZhangJianyu — COMMENTED** [review]
It's great to involve the oneDNN in SYCL backend.

1. If oneDNN is detected, oneMKL won't be used in whole code.
Is it right?

2. Please update the SYCL.md with oneMKL and oneDNN dependence info.

**Comment by luoyu-intel** [comment]
> It's great to involve the oneDNN in SYCL backend.
> 
> 1. If oneDNN is detected, oneMKL won't be used in whole code.
>    Is it right?
> 2. Please update the SYCL.md with oneMKL and oneDNN dependence info.

1. Yes.
2. MD has been updated.

**Comment by AGenchev** [comment]
Interesting. I wonder whether this thing, similar to OneDNN https://github.com/amd/ZenDNN could be integrated into  SYCL backend.

**Comment by vpirogov** [comment]
> Interesting. I wonder whether this thing, similar to OneDNN https://github.com/amd/ZenDNN could be integrated into SYCL backend.

This thing is CPU only and is not intended to be used as standalone software. oneDNN will cover AMD client and server CPUs nicely if necessary.

**Pull request merged**